### DISCUSSION TITLE
Overlap of return annotation with `[source]`

### DISCUSSION THREAD

**Discussion by harshil21** [pradyunsg/furo#350] [Potential Issue] [answered]
### What's happening?

When using `linkcode` to attach the `[source]` for each method, and when using `autodoc-process-signature` to manually generate a custom return annotation, there is an UI overlap of both those elements.

![image](https://user-images.githubusercontent.com/37377066/148620602-7c519ad3-1ef1-4831-83df-3f2c7a72a2cc.png)

My `html_theme_options` (unrelated to above but still):
``` python
html_theme_options = {
    # 'style_external_links': True,  # Use after https://github.com/pradyunsg/furo/discussions/348
    'navigation_with_keys': True,
    'dark_css_variables': {'admonition-title-font-size': '0.95rem',
                           'admonition-font-size': '0.92rem'},
    'light_css_variables': {'admonition-title-font-size': '0.95rem',
                            'admonition-font-size': '0.92rem'}
}
```
`autodoc_type_hints` is set to `none`.

Furo version: 2022.1.2
Sphinx version: 4.3.2

### Reproducer

1. Clone `https://github.com/harshil21/python-telegram-bot.git`
2. Switch to the `dark-mode-adjust` branch (`git checkout dark-mode-adjust`)
3. Run the command `make -C docs html`
4. Now after the docs are generated, open `docs/build/html/telegram.bot.html` in a browser
5. Search or scroll down to the method `forwardMessage`
6. See the UI bug.

----
Another place where it happens with the type annotation of the method:
See (https://python-telegram-bot.readthedocs.io/en/doc-fixes/telegram.ext.updater.html#telegram.ext.Updater.idle)
1. Clone https://github.com/python-telegram-bot/python-telegram-bot.git
2. Follow above steps
4. Open `docs/build/html/telegram.ext.updater.html` in a browser
5. Search or scroll down to the method `idle`
6. See the same UI bug

### Expectation

Furo should recognize that there will be an overlap and put it in the next line, like it did below:
![image](https://user-images.githubusercontent.com/37377066/148621082-c7c6d1b0-346b-4e82-9b39-bde8d56b823d.png)

### Code of Conduct

- [X] I agree to follow the [Code of Conduct](https://github.com/pradyunsg/furo/blob/main/CODE_OF_CONDUCT.md).

**Comment by Bibo-Joshi** [comment]
IISc with `autodoc_type_hints=None`, the return type shouldn't show at all, right?

**Reply by harshil21** [reply]
right, I was testing showing only the return annotation using `autodoc-process-signature`

**Comment by pradyunsg** [comment]
> Will commit changes to remote soon, so I'll edit this then

There's nothing actionable here for me until then. I won't be able to change things unless I'm able to reproduce this.

**Reply by harshil21** [reply]
I've edited the OP with instructions to reproduce :)

**Comment by francois-rozet** [comment]
Hello @pradyunsg,

On the same subject, if the last line overlaps with the `[source]`, the latter is forced out of the signature which is quite ugly.

<img width="565" alt="image" src="https://user-images.githubusercontent.com/37352336/161766249-b8c7a56f-4f63-45af-8cc8-cca88c4dcc67.png">

I tried to come up with a CSS-only fix, but it never worked because the permalink tag comes **after** the `[source]` link in the HTML.

The solution I have currently consists in 
1. swapping the two elements (`a.reference <-> a.headerlink`) with `sed` after the HTML is generated and
2. modifying the CSS such that the permalink cannot overlap with the `[source]`.

    ```css
    .sig > a.headerlink:not(:last-child) {
        margin-right: 5em;
    }

    .sig > a.reference {
        float: right;
        width: 0;
    }
    ```

<img width="574" alt="image" src="https://user-images.githubusercontent.com/37352336/161771176-d44238e1-80aa-48a4-bc07-3c2d11817c56.png">

Do you think it would be possible to add the "swap" in Furo ?

**Comment by pradyunsg** [comment, marked as answer]
I investigated this today, and managed to reproduce this. c718f861a289b53b1f454ad3818c7a195760a280 makes this work correctly and should be doing what we want here. :)

**Reply by francois-rozet** [reply]
Hello @pradyunsg, I am not sure that your fix works. Just adding two more `a`s makes the `[source]` link overflow out of the signature.

![image](https://github.com/pradyunsg/furo/assets/37352336/a1ab3c4f-76df-4020-9897-a0595bc5e913)

As mentioned in a previous comment, the only reliable fix I found was to swap the order of the permalink (`#`) and source link in HTML.

**Reply by pradyunsg** [reply]
That markup comes directly from Sphinx, and I've in-general avoided doing HTML transformation for things like that in this theme.
